pkg/glusterfs/Dockerfile

```diff
@@ -1,22 +1,71 @@
 # Copyright 2018 The Gluster CSI Authors.
 
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-#You may obtain a copy of the License at
-
-#    http://www.apache.org/licenses/LICENSE-2.0
+# Licensed under GNU LESSER GENERAL PUBLIC LICENSE Version 3, 29 June 2007
+# You may obtain a copy of the License at
+#    https://opensource.org/licenses/lgpl-3.0.html
 
 # Unless required by applicable law or agreed to in writing, software
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM centos:7.5.1804
 
-# Copy glusterfs-csi-driver from build directory
-COPY ./build/glusterfs-csi-driver /glusterfs-csi-driver
+# Establish the base image that will be used
+ARG BASE_IMAGE="centos:7.5.1804"
+
+#-- Build phase
+FROM ${BASE_IMAGE} AS build
+
+ENV GOPATH="/go/" \
+    SRCDIR="/go/src/github.com/gluster/gluster-csi-driver/"
+
+RUN yum install -y \
+    git \
+    golang
+
+# Install dep
+RUN mkdir -p /go/bin
+RUN curl -L https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
+
+# Vendor dependencies
+COPY Gopkg.lock Gopkg.toml "${SRCDIR}"
+WORKDIR "${SRCDIR}"
+RUN /go/bin/dep ensure -v -vendor-only
+
+# Build executable
+COPY . "${SRCDIR}"
```

Comment:
will this copy everything including `Gopkg.lock and Gopkg.toml`?
if yes we can move this to line 32

Reply:
Yes, it will re-copy `Gopkg.*`, but...
The reason I don't do the full copy on L32 is to minimize the invalidations of the docker image cache. `dep ensure` is a very time-consuming step, so it's best to maximize hits in the image cache. Since changes to any `COPY`-ed file will invalidate the cache, if we were to copy all the files on L32, a change to any file would cause the vendor directory to be re-downloaded. Instead, I only add the 2 files necessary (`Gopkg.*`) prior to the `dep ensure`.

Reply:
I'm not going to hold up the PR on this, but I just want clarity on something: You say you run `dep ensure` before the source COPY so as to minimize cache invalidation. However, I don't see any modifications to/in `SRCDIR` after the source copy, so I don't understand how your workaround prevents cache invalidations, as opposed to just copying the entire source and running dep ensure afterwards.

Reply:
Bingo! 😀